### PR TITLE
New configuration for larger text columns, Adding metadata columns to Jdbc outputs, General changes in organization; Cortina CDA changes

### PR DESCRIPTION
**Pull request by scottrhodes**
Changes in prior PR, plus:

Made large text field list configurable in config.yaml.
Organized adding and removing metadata columns to data frames dependent on write type.
Added gwcbi___ metadata columns to the Raw Jdbc write. Also added custom gwcdac__fingerprintfolder and gwcdac__timestampfolder metadata columns to Raw Jdbc write.

Readme.MD updated

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,7 +127,15 @@ Under "*outputLocation*":
 * "*includeColumnNames*" boolean (defaults to false) - should be "true" to include a row of column names at the top of the csv file for each table, and "false" for no header row.
 * "*saveAsSingleFile*" boolean (defaults to false) - should be "true" for writing out a single file (.csv or .parquet) per table, and "false" to have multiple/smaller files be written based on SPARK partitioning.
 * "*saveIntoTimestampDirectory*" boolean (defaults to false) - should be "true" to save the CSV files into a directory with savepoint timestamp (/outputLocation/path/table/timestamp/*.csv), and "false" to save directly into the table directory (/outputLocation/path/table/*.csv).
-
+* "*largeTextFields*" is a comma delimited list of table.column values that must allow max length varchar types.
```
Comment (baby-bell) [inline README.md]: This section will conflict with #2, so you should talk to Sasi before merging (or she should talk to you, whatever works).

File: build.gradle [hunk 2]
```diff
@@ -30,8 +30,8 @@ dependencies {
   implementation "com.amazonaws:aws-java-sdk-s3:$awsSDKVersion"
   implementation "org.apache.commons:commons-lang3:$apacheCommonsVersion"
   implementation "org.rogach:scallop_$scalaVersion:$scallopVersion"
-  implementation "com.microsoft.sqlserver:mssql-jdbc:$sqlserverVersion"
-  implementation "org.postgresql:postgresql:$postgresqlVersion"
+  compile "com.microsoft.sqlserver:mssql-jdbc:$sqlserverVersion"
```
Comment (baby-bell) [inline build.gradle:33]: Why is this necessary? I believe that `compile` is deprecated.

File: src/main/scala/com/guidewire/cda/TableReader.scala [hunk 3]
```diff
@@ -464,16 +466,42 @@ class TableReader(clientConfig: ClientConfig) {
    */
   private[cda] def fetchDataFrameForTableTimestampSubfolder(tableTimestampSubfolderInfo: TableS3LocationWithTimestampInfo): DataFrameWrapper = {
     val s3URI = tableTimestampSubfolderInfo.timestampSubfolderURI
-    val s3aURL = s"s3a://${s3URI.getBucket}/${s3URI.getKey}*"
+    val s3aURL = s"s3a://${s3URI.getBucket}/${s3URI.getKey}*.parquet"
```
Comment (baby-bell) [inline src/main/scala/com/guidewire/cda/TableReader.scala:469]: 👍

File: src/main/scala/com/guidewire/cda/TableReader.scala [hunk 4]
```diff
@@ -464,16 +466,42 @@ class TableReader(clientConfig: ClientConfig) {
...
-    DataFrameWrapper(tableTimestampSubfolderInfo.tableName, tableTimestampSubfolderInfo.schemaFingerprint,
-      dataFrameNoInternalColumns)
+    val dataFrameAdditionalColumns = addCdaClientColumns(dataFrame, tableTimestampSubfolderInfo)
+
+    clientConfig.outputSettings.exportTarget match {
+      case "file" =>
+        DataFrameWrapper(tableTimestampSubfolderInfo.tableName, tableTimestampSubfolderInfo.schemaFingerprint, dataFrameNoInternalColumns)
+      case "jdbc" => {
+        // Remove columns that will not insert properly into database tables ("spatial", "textdata")
+        val dropList = dataFrameAdditionalColumns.columns.filter(colName => colName.toLowerCase.contains("spatial") || colName.toLowerCase.equals("textdata"))
```
Comment (baby-bell) [inline src/main/scala/com/guidewire/cda/TableReader.scala:495]: I would strongly suggest adding a note about this to the README

File: build.gradle (hunk 2 above)
Comment (scottrhodes) [reply build.gradle:33]: What we found is that if you leave it as "implementation" the code doesn't pick up the correct driver at run time - it has to be installed separately and managed. Especially true with PostgreSQL.  Changing to "compile" fixes that problem.
I have warnings for deprecated items turned on in IntelliJ - it isn't reporting any issues.

File: src/main/scala/com/guidewire/cda/TableReader.scala [hunk 5]
```diff
@@ -484,13 +512,22 @@ class TableReader(clientConfig: ClientConfig) {
...
-    val dropList = dataFrame.columns.filter(colName => (colName.toLowerCase.startsWith("gwcbi___") && !relevantInternalColumns.contains(colName))
-      // Get rid of geospatial columns.
-      || colName.toLowerCase.contains("spatial")
-      // Get rid of textdata columns. Contains XML data that won't insert into the database properly.
-      || colName.toLowerCase.equals("textdata"))
+    val dropList = dataFrame.columns.filter(colName => (colName.toLowerCase.startsWith("gwcbi___") && !relevantInternalColumns.contains(colName)))
     dataFrame.drop(dropList: _*)
+  }
 
+  /** Function that takes a DataFrame corresponding to some table and drops internal columns that
```
Comment (baby-bell) [inline src/main/scala/com/guidewire/cda/TableReader.scala]: This documentation is factually incorrect. Copy-paste error?

File: src/main/scala/com/guidewire/cda/config/ClientConfigReader.scala [hunk 6]
```diff
@@ -46,6 +46,8 @@ private[cda] case class JdbcConnectionMerged(jdbcUsername: String,
                                              jdbcSchema: String,
                                              jdbcApplyLatestUpdatesOnly: Boolean)
 
+private[cda] case class KafkaSettings(bootstrapServer: String)
```
Comment (baby-bell) [inline src/main/scala/com/guidewire/cda/config/ClientConfigReader.scala]: Kafka settings? This looks like it was left in accidentally.

File: README.md (hunk 1 above)
Comment (scottrhodes) [reply README.md]: Yes - I've made a few changes to the code since the #2 PR. I'll look at that version and merge them in my master branch (the PR source). I'll reach out to Sasi if I had questions.
I should be able to get to that today.

File: src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala [hunk 7]
```diff
@@ -36,11 +36,14 @@ import org.apache.spark.sql.functions.col
...
 import org.apache.spark.sql.types.StructType
 
-
 private[outputwriter] class JdbcOutputWriter(override val outputPath: String, override val includeColumnNames: Boolean,
                                              override val saveAsSingleFile: Boolean, override val saveIntoTimestampDirectory: Boolean,
                                              override val clientConfig: ClientConfig) extends OutputWriter {
 
+  val tableDotColumnValues: Array[String] = Option(clientConfig.outputSettings.largeTextFields).getOrElse("").replace(" ", "").split(",")
+
+  private[outputwriter] val tableColumnValues:Array[String] = tableDotColumnValues
```
Comment (baby-bell) [inline src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala]:
```suggestion
  private[outputwriter] val tableColumnValues: Set[String] = Option(clientConfig.outputSettings.largeTextFields).getOrElse("").replace(" ", "").split(",").toSet
```
You're checking whether an element is contained within it, make it a set.

Comment (baby-bell) [reply src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala]: Also, I would rename to something that makes the purpose apparent. I don't know what `tableColumnValues` is supposed to refer to.

File: src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala [hunk 8]
```diff
@@ -326,10 +329,11 @@ private[outputwriter] class JdbcOutputWriter(override val outputPath: String, ov
       }
 
       // Build the sql Update statement to be used as a prepared statement for the Updates.
-      val colListForSetClause = latestChangeForEachID.columns.filter(_ != "id")
+      val colListForSetClause = latestChangeForEachID.columns
+        .filter(_ != "id")
+        .filter(_ != "hexseqvalue")
```
Comment (baby-bell) [inline src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala:332]:
```suggestion
        .filter(_ != "id" && _ != "hexseqval" )
```

File: src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala [hunk 9]
```diff
@@ -572,18 +576,14 @@ private[outputwriter] class JdbcOutputWriter(override val outputPath: String, ov
...
       val tableNameNoSchema = tableName.substring(tableName.indexOf(".") + 1)
-      (tableNameNoSchema, fieldName) match {
-        case ("cc_outboundrecord", "\"content\"") |
-             ("cc_contactorigvalue", "\"origvalue\"") |
-             ("pc_diagratingworksheet", "\"diagnosticcapture\"") |
-             ("cc_note", "\"body\"") => largeStringDataType
-        case _                       => stringDataType
-      }
+      val currentTableColumn = (tableNameNoSchema+"."+fieldName)
+      if (!tableDotColumnValues.filter(_ == currentTableColumn).isEmpty) largeStringDataType
```
Comment (baby-bell) [inline src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala]:
```suggestion
      if (tableColumnValues.contains(currentTableColumn)) largeStringDataType
```

File: sample.config.yaml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  ###ORACLE###
+  #  jdbcUsername: <user>
+  #  jdbcPassword: <password>
+  #  jdbcUrl: 'jdbc:oracle:thin:XEPDB1/xepdb1@//localhost:1521/XEPDB1'
+  #  jdbcSchema: '<schema>'
+
+  ###MSSQL###
+  jdbcUsername: novalue
+  jdbcPassword: nopwd
+  jdbcUrl: 'jdbc:sqlserver://<host>\<instance>;databaseName=<database name>;integratedSecurity=true'
```
Comment (baby-bell) [inline sample.config.yaml:44]:
```suggestion
  jdbcUrl: 'jdbc:sqlserver://<host>\\<instance>;databaseName=<database name>;integratedSecurity=true'
```
backslash should be escaped

File: sample.config.yaml [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
```
Comment (baby-bell) [inline sample.config.yaml:65]:
```suggestion
  jdbcUrl: 'jdbc:sqlserver://<host>\\<instance>;databaseName=<database name>;integratedSecurity=true'
```
ditto

File: src/main/scala/com/guidewire/cda/TableReader.scala [hunk 12]
```diff
@@ -464,16 +466,42 @@ class TableReader(clientConfig: ClientConfig) {
...
-    //   double quotes around it in the table definition. When we execute the Merged code, it thinks the value is a literal and won't insert.
-    //   Adding additional .withColumnRenamed() lines below works in the event multiple columns need to be renamed.
     val dataFrame = spark.sqlContext.read.parquet(s3aURL)
-      .withColumnRenamed("interval", "interval_")
+
+    val dataFrameWrapper = manageDataFrameColumns(dataFrame, tableTimestampSubfolderInfo)
+    dataFrameWrapper
+  }
+
+  /** Function that takes a DataFrame corresponding to some table and manages the addition and removal of
```
Comment (baby-bell) [inline src/main/scala/com/guidewire/cda/TableReader.scala:477]: This documentation does not mention dropping of `spatial`/`textdata` fields. Are these internal fields, or just ones that the database cannot handle?

**Review by baby-bell — CHANGES_REQUESTED** [review]
Some of the documentation is factually incorrect. Also some minor code changes. Otherwise LGTM.

Side note: I notice a lot of use of braceless if-else. Would be nice to avoid that; not specific to this PR.

File: build.gradle (hunk 2 above)
Comment (baby-bell) [reply build.gradle:33]: Compile is deprecated in Gradle 6: https://github.com/gradle/gradle/issues/15661
IntelliJ won't report that since we are using Gradle 5. We can resolve this later, not the highest priority issue.

File: src/main/scala/com/guidewire/cda/TableReader.scala (hunk 5 above)
Comment (scottrhodes) [reply src/main/scala/com/guidewire/cda/TableReader.scala]: I'll look at it

File: src/main/scala/com/guidewire/cda/config/ClientConfigReader.scala (hunk 6 above)
Comment (scottrhodes) [reply src/main/scala/com/guidewire/cda/config/ClientConfigReader.scala]: It was left in accidentally. I'll scrub everything Kafka

File: sample.config.yaml (hunk 11 above)
Comment (scottrhodes) [reply sample.config.yaml:65]: Code has been working in the current state for months - haven't had to escape those backslashes. Only applies to SQL Server connection strings when there are multiple instances of SQL Server on one host

File: src/main/scala/com/guidewire/cda/TableReader.scala (hunk 12 above)
Comment (scottrhodes) [reply src/main/scala/com/guidewire/cda/TableReader.scala:477]: These are fields the database cannot handle.

Spatial is actually an array of two elements making up the geospatial information - I could probably make the code handle it but just haven't done so.
I need to review the textdata field again - early on the code choked on it so we set it aside. I'll revisit.

File: sample.config.yaml (hunk 11 above)
Comment (baby-bell) [reply sample.config.yaml:65]: OK. If it ain't broke...

File: src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala (hunk 8 above)
Comment (scottrhodes) [reply src/main/scala/gw/cda/api/outputwriter/JdbcOutputWriter.scala:332]: Won't resolve the second '!='. May be why I had them on separate lines.
image

leaving as is for now. will dig into it later.

File: README.md (hunk 1 above)
Comment (scottrhodes) [reply README.md]: Worked with Sasi - finalizing changes across both PRs with her.

**Comment by gw-vsasidharan** [comment]
LGTM!

**Comment by sh-praveen** [comment]
@gw-vsasidharan @scottrhodes @baby-bell @malyw @benjaminm3 
Need your help!
Do we have the **sample parquet files** checked in? 
If not, can I find it in any external location? Please suggest.
The manifest and savepoints files refer to a series of tables, for which the parquet files are expected in S3 bucket.

**Comment by baby-bell** [comment]
@sh-praveen , please file an issue
